File: src/so/aporia/module/ModuleManager.java
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.module;

import so.aporia.module.impl.combat.Aura;
import so.aporia.module.impl.misc.AutoConfig;
import so.aporia.module.impl.misc.AutoEZ;
import so.aporia.module.impl.misc.ClickGui;
import so.aporia.module.impl.misc.DiscordRPCModule;
import so.aporia.module.impl.misc.PacketDebug;
import so.aporia.module.impl.misc.ServerHelper;
import so.aporia.module.impl.misc.TestModule;
import so.aporia.module.impl.move.AutoSprint;
import so.aporia.module.impl.render.Beautifully;
import so.aporia.module.impl.render.Hud;
import so.aporia.module.impl.render.PlayerESP;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

/**
 * Central registry for all modules.
 * Handles registration, keybind dispatch, and lifecycle.
 */
public final class ModuleManager {

    public static final ModuleManager INSTANCE = new ModuleManager();

    private final List<Module> modules = new ArrayList<>();

    private ModuleManager() {
        registerAll(
                // render
            new Hud(),
            new Beautifully(),
            new PlayerESP(),
            new ClickGui(),
                // misc
            new ServerHelper(),
            new AutoConfig(),
            new AutoEZ(),
            new TestModule(),
            new PacketDebug(),
            new DiscordRPCModule(),
                // move
            new AutoSprint(),
                // combat
            new Aura()
        );
    }

    public void register(Module module) {
        modules.add(module);
    }

    public void registerAll(Module... mods) {
        for (Module m : mods) register(m);
    }

    public Collection<Module> getAll() {
        return Collections.unmodifiableList(modules);
    }

    public List<Module> getByCategory(Category category) {
        List<Module> out = new ArrayList<>();
        for (Module m : modules) if (m.category() == category) out.add(m);
        return out;
    }

    public Module get(String name) {
        for (Module m : modules) if (m.name().equalsIgnoreCase(name)) return m;
        return null;
    }
}

File: src/so/aporia/module/ModuleManager.kt
package so.aporia.module
import com.chaos.annotation.ChaosNative
@ChaosNative
object ModuleManager {

    private val modules = mutableListOf<Module>()
    private val scriptModules = mutableListOf<Module>()

    init {
        registerAllDefaultModules()
        registerScriptModules()
    }

    private fun registerAllDefaultModules() {
        registerAll(
            so.aporia.module.impl.render.Hud(),
            so.aporia.module.impl.render.Beautifully(),

            so.aporia.module.impl.render.NameTags(),
            so.aporia.module.impl.misc.ClickGui(),
            so.aporia.module.impl.misc.ServerHelper(),
            so.aporia.module.impl.misc.AutoConfig(),
            so.aporia.module.impl.misc.AutoEZ(),
            so.aporia.module.impl.misc.TestModule(),
            so.aporia.module.impl.misc.PacketDebug(),
            so.aporia.module.impl.misc.Disabler(),
            so.aporia.module.impl.misc.DiscordRPCModule(),
            so.aporia.module.impl.combat.AutoGapple(),
            so.aporia.module.impl.combat.AutoTotem(),
            so.aporia.module.impl.combat.God(),
            so.aporia.module.impl.player.NoPush(),
            so.aporia.module.impl.player.ElytraHelper(),
            so.aporia.module.impl.world.MiddleClick(),
            so.aporia.module.impl.world.FakeLag(),
            so.aporia.module.impl.move.Speed(),
            so.aporia.module.impl.move.AutoSprint(),
            so.aporia.module.impl.move.Velocity(),
            so.aporia.module.impl.move.Flight(),
            so.aporia.module.impl.combat.Criticals(),
            so.aporia.module.impl.combat.TPAura(),
            so.aporia.module.impl.combat.MaceExploit(),
            so.aporia.module.impl.combat.Aura(),
            so.aporia.module.impl.combat.SpearTarget(),
            so.aporia.module.impl.combat.NoFriendDamage(),
            so.aporia.module.impl.combat.ElytraTarget(),
            so.aporia.module.impl.render.NoRender(),
            so.aporia.module.impl.render.PlayerESP(),
            so.aporia.module.impl.render.Hands(),
            so.aporia.module.impl.render.WorldRenderer()
        )
    }

    private fun registerScriptModules() {
        registerScript(so.aporia.module.impl.render.clickgui.ThemeManagerModule())
    }

    @JvmStatic
    fun register(module: Module) {
        modules.add(module)
    }

    @JvmStatic
    fun registerAll(vararg mods: Module) {
        for (m in mods) register(m)
    }

    @JvmStatic
    fun registerScript(module: Module) {
        modules.add(module)
        scriptModules.add(module)
    }

    @JvmStatic
    fun getAll(): List<Module> = modules.toList()

    @JvmStatic
    fun getScriptModules(): List<Module> = scriptModules.toList()

    @JvmStatic
    fun getByCategory(category: Category): List<Module> {
        return modules.filter { it.category == category && it !in scriptModules }
    }

    @JvmStatic
    fun get(name: String): Module? {
        return modules.firstOrNull { it.name.equals(name, ignoreCase = true) }
    }
}